# NLST Predictions Merge

Match and put together NLST predictions for Kiran, Thijmen, PanCan and Sybil (after those predictions have ran).

In [1]:
import pandas as pd
import os
import numpy as np

import seaborn as sns
sns.set_style("white")
from evalutils.roc import get_bootstrapped_roc_ci_curves
import matplotlib.pyplot as plt

%matplotlib inline
import sklearn.metrics as skl_metrics

## directory where results are
LOCAL_PC = False
root_dir = "/mnt/w" if LOCAL_PC else "/data/bodyct"
EXPERIMENT_DIR = f"{root_dir}/experiments/lung-malignancy-fairness-shaurya"
NLST_PREDS = f"{EXPERIMENT_DIR}/nlst-preds"

NLST_PREDS_LOCAL = "/mnt/c/Users/shaur/OneDrive - Radboudumc/Documents - Master - Shaurya Gaur/General/Malignancy-Estimation Results/nlst"
NLST_PREDS = NLST_PREDS_LOCAL ## Comment out if not using Teams backup (aka Chansey is up :)

## Load Predictions

Kiran's model and PanCan2b.

In [2]:
nlst_kiran_pancan = pd.read_csv(f"{NLST_PREDS}/NLST_DL_vs_PanCan_Venk21.csv")
nlst_kiran_pancan.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16077 entries, 0 to 16076
Data columns (total 25 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   PatientID            16077 non-null  int64  
 1   StudyDate            16077 non-null  int64  
 2   SeriesInstanceUID    16077 non-null  object 
 3   CoordX               16077 non-null  float64
 4   CoordY               16077 non-null  float64
 5   CoordZ               16077 non-null  float64
 6   LesionID             16077 non-null  int64  
 7   NoduleType           16077 non-null  object 
 8   Spiculation          16077 non-null  bool   
 9   Diameter [mm]        16077 non-null  float64
 10  Age                  16077 non-null  int64  
 11  Gender               16077 non-null  int64  
 12  FamilyHistoryLungCa  16077 non-null  bool   
 13  Emphysema            16077 non-null  bool   
 14  NoduleInUpperLung    16077 non-null  bool   
 15  NoduleCounts         16077 non-null 

Thijmen's model. Note that it only has 3240 annotation predictions and not 16077.

In [4]:
nlst_thijmen = pd.read_csv(f"{NLST_PREDS}/NLST_Tijmen_results/merged_model_output.csv")
nlst_thijmen.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3240 entries, 0 to 3239
Data columns (total 3 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   AnnotationID  3240 non-null   object 
 1   label         3240 non-null   float64
 2   Thijmen_mean  3240 non-null   float64
dtypes: float64(2), object(1)
memory usage: 76.1+ KB


Sanity check to make sure the labels match for the `AnnotationID` between Thijmen's CSV and the Kiran PanCan sheet. If the output DataFrame below is empty, you're all good :)

In [17]:
nlst_kp2 = nlst_kiran_pancan.sort_values(by='AnnotationID', ascending=True).set_index("AnnotationID")
nlst_tj2 = nlst_thijmen.sort_values(by='AnnotationID', ascending=True).set_index("AnnotationID")

nlst_kp2 = nlst_kp2[nlst_kp2.index.isin(nlst_tj2.index)]
res = nlst_kp2['label'].compare(nlst_tj2['label'])

del nlst_kp2, nlst_tj2
res

Empty DataFrame
Columns: [self, other]
Index: []

Merge the predictions!

In [23]:
nlst_preds = nlst_kiran_pancan.merge(nlst_thijmen, 
                                     how="left", ## 'inner' for only ones Thijmen has preds for
                                     on=['AnnotationID', 'label'], suffixes=(None,None))
nlst_preds.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16077 entries, 0 to 16076
Data columns (total 26 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   PatientID            16077 non-null  int64  
 1   StudyDate            16077 non-null  int64  
 2   SeriesInstanceUID    16077 non-null  object 
 3   CoordX               16077 non-null  float64
 4   CoordY               16077 non-null  float64
 5   CoordZ               16077 non-null  float64
 6   LesionID             16077 non-null  int64  
 7   NoduleType           16077 non-null  object 
 8   Spiculation          16077 non-null  bool   
 9   Diameter [mm]        16077 non-null  float64
 10  Age                  16077 non-null  int64  
 11  Gender               16077 non-null  int64  
 12  FamilyHistoryLungCa  16077 non-null  bool   
 13  Emphysema            16077 non-null  bool   
 14  NoduleInUpperLung    16077 non-null  bool   
 15  NoduleCounts         16077 non-null 

Thijmen's doesn't seem like a size-matched subset with a higher malignant percentage.

In [26]:
nlst_thijmen_nonnull = nlst_preds[~nlst_preds['Thijmen_mean'].isnull()]
nlst_thijmen_nonnull.groupby('label').nunique()

PatientID  StudyDate  SeriesInstanceUID  CoordX  CoordY  CoordZ  \
label                                                                    
0            907          3               1797    2794    2834    2845   
1            154          3                265     279     280     280   

       LesionID  NoduleType  Spiculation  Diameter [mm]  ...  NoduleCounts  \
label                                                    ...                 
0            18           4            1            202  ...            10   
1             2           3            2            159  ...             8   

       SCT_EPI_LOC  xie_gc_gclobe150  loclup  locrup  PanCan2b    DL  \
label                                                                  
0                7                 0       0       0      2722  2958   
1                6                 6       2       2       280   281   

       NoduleID  AnnotationID  Thijmen_mean  
label                                        
0          1825          2959          2959  
1           164           281           281  

[2 rows x 25 columns]

## Load NLST Demographic Info

In [139]:
nlst_dict_full = pd.read_csv(f"{NLST_PREDS}/participant_d040722.csv")
nlst_dict_full.info()

/tmp/ipykernel_2342/2896154133.py:1: DtypeWarning: Columns (239,240,348) have mixed types. Specify dtype option on import or set low_memory=False.
  nlst_dict_full = pd.read_csv(f"{NLST_PREDS}/participant_d040722.csv")


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 53452 entries, 0 to 53451
Columns: 356 entries, cen to evpsent
dtypes: float64(283), int64(64), object(9)
memory usage: 145.2+ MB


In [140]:
nlst_dict_subset = nlst_dict_full[nlst_dict_full['pid'].isin(nlst_preds['PatientID'])].reset_index(drop=True)
nlst_dict_subset.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5282 entries, 0 to 5281
Columns: 356 entries, cen to evpsent
dtypes: float64(283), int64(64), object(9)
memory usage: 14.3+ MB


In [141]:
nlst_dict_subset.groupby(['gender'])['anyscr_has_nodule'].value_counts(dropna=False, normalize=True)

gender  anyscr_has_nodule
1       1.0                  0.986711
        0.0                  0.013289
2       1.0                  0.994278
        0.0                  0.005722
Name: proportion, dtype: float64

In [147]:
workfacets = ['res', 'wrk', 'yrs']
worktypes = ['asbe', 'baki', 'butc', 'chem', 'coal', 'cott', 'farm', 'fire', 'flou', 'foun', 'hard', 'pain', 'sand', 'weld']

diseasefacets = ['age', 'diag']
diseasetypes = ['adas', 'asbe', 'bron', 'chas', 'chro', 'copd', 'diab', 'emph', 'fibr', 'hear', 'hype', 'pneu', 'sarc', 'sili', 'stro', 'tube']

pcancerfacets = ['age', 'canc']
pcancertypes = ['blad', 'brea', 'cerv', 'colo', 'esop', 'kidn', 'lary', 'lung', 'nasa', 'oral', 'panc', 'phar', 'stom', 'thyr', 'tran']

alcohol_acrin = ['acrin_alc_curr', 'acrin_alc_ever', 'acrin_drink24hr', 'acrin_drinknum_curr', 'acrin_drinknum_form', 'acrin_drinkyrs_curr', 'acrin_drinkyrs_form', 'acrin_lastdrink']
alcohol_lss = ['lss_alcohol_freq', 'lss_alcohol_num']

cols = {
    "info": ['pid', 'study'],
    "demo": ['age', 'educat', 'ethnic', 'gender', 'height', 'marital', 'race', 'weight'],
    "smoking": ['age_quit', 'cigar', 'cigsmok', 'pipe', 'pkyr', 'smokeage', 'smokeday', 'smokelive', 'smokework', 'smokeyr'],
    "screen": ['scr_group'],
    "work": [facet+job for facet in workfacets for job in worktypes],
    "disease": [facet+disease for facet in diseasefacets for disease in diseasetypes],
    "personal_cancer": [facet+cancer for facet in pcancerfacets for cancer in pcancertypes],
    "family": ['fam'+member for member in ['brother', 'sister', 'child', 'father', 'mother']],
    "alcohol": alcohol_lss+alcohol_acrin,   
}

allcols = sum([cols[k] for k in cols], start=[])
len(allcols)

140

In [152]:
for c in cols:
    print(c, len(cols[c]))
    nlst_colinfo = nlst_dict_subset[cols[c]]
    display(nlst_colinfo.describe())

info 2


pid        study
count    5282.000000  5282.000000
mean   148184.307081     1.474252
std     44395.945951     0.728020
min    100012.000000     1.000000
25%    113113.500000     1.000000
50%    125989.000000     1.000000
75%    205225.250000     2.000000
max    218887.000000     3.000000

demo 8


age       educat       ethnic       gender       height  \
count  5282.000000  5282.000000  5282.000000  5282.000000  5273.000000   
mean     62.183453     4.835290     2.538054     1.430140    67.846008   
std       5.184738     4.579593     7.283255     0.495142     3.921241   
min      54.000000     1.000000     1.000000     1.000000    49.000000   
25%      58.000000     3.000000     2.000000     1.000000    65.000000   
50%      61.000000     5.000000     2.000000     1.000000    68.000000   
75%      66.000000     6.000000     2.000000     2.000000    71.000000   
max      74.000000    99.000000    99.000000     2.000000    83.000000   

           marital         race       weight  
count  5281.000000  5282.000000  5270.000000  
mean      2.654043     1.307459   181.264706  
std       1.247796     3.831637    38.917809  
min       1.000000     1.000000    75.000000  
25%       2.000000     1.000000   154.000000  
50%       2.000000     1.000000   179.000000  
75%       3.000000     1.000000   205.000000  
max       9.000000    99.000000   368.000000

smoking 10


age_quit        cigar      cigsmok         pipe         pkyr  \
count  2618.000000  5264.000000  5282.000000  5247.000000  5282.000000   
mean     55.744843     0.194149     0.503786     0.219173    58.086814   
std       6.603271     0.395582     0.500033     0.413725    24.839586   
min      11.000000     0.000000     0.000000     0.000000    30.000000   
25%      51.000000     0.000000     0.000000     0.000000    40.500000   
50%      56.000000     0.000000     1.000000     0.000000    50.000000   
75%      60.000000     0.000000     1.000000     0.000000    69.000000   
max      74.000000     1.000000     1.000000     1.000000   260.000000   

          smokeage     smokeday    smokelive    smokework      smokeyr  
count  5282.000000  5282.000000  5259.000000  5247.000000  5282.000000  
mean     16.585384    28.639909     0.883248     0.871355    40.993942  
std       3.551498    11.542460     0.321155     0.334838     7.403968  
min       4.000000    12.000000     0.000000     0.000000    13.000000  
25%      15.000000    20.000000     1.000000     1.000000    36.000000  
50%      16.000000    25.000000     1.000000     1.000000    41.000000  
75%      18.000000    35.000000     1.000000     1.000000    46.000000  
max      40.000000   135.000000     1.000000     1.000000    67.000000

screen 1


scr_group
count  5282.000000
mean      1.962514
std       0.559739
min       1.000000
25%       2.000000
50%       2.000000
75%       2.000000
max       5.000000

work 42


resasbe     resbaki     resbutc     reschem    rescoal    rescott  \
count  246.000000  132.000000  109.000000  333.000000  26.000000  41.000000   
mean     0.186992    0.007576    0.018349    0.183183   0.230769   0.048780   
std      0.439844    0.087039    0.134829    0.424498   0.587040   0.218085   
min      0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
25%      0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
50%      0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
75%      0.000000    0.000000    0.000000    0.000000   0.000000   0.000000   
max      2.000000    1.000000    1.000000    2.000000   2.000000   1.000000   

          resfarm    resfire    resflou     resfoun  ...    yrscoal  \
count  566.000000  93.000000  53.000000  219.000000  ...  26.000000   
mean     0.033569   0.548387   0.132075    0.146119  ...  16.076923   
std      0.198945   0.561754   0.394078    0.390977  ...  13.962587   
min      0.000000   0.000000   0.000000    0.000000  ...   2.000000   
25%      0.000000   0.000000   0.000000    0.000000  ...   2.250000   
50%      0.000000   1.000000   0.000000    0.000000  ...  16.500000   
75%      0.000000   1.000000   0.000000    0.000000  ...  24.500000   
max      2.000000   2.000000   2.000000    2.000000  ...  55.000000   

         yrscott     yrsfarm    yrsfire    yrsflou     yrsfoun    yrshard  \
count  42.000000  578.000000  93.000000  54.000000  222.000000  36.000000   
mean    7.571429   13.043253  14.752688   9.148148   13.909910  10.277778   
std     9.625025   10.545430  10.478400  11.296029   13.636556  10.188540   
min     1.000000    1.000000   1.000000   1.000000    1.000000   1.000000   
25%     1.000000    5.000000   5.000000   2.000000    3.000000   2.000000   
50%     4.000000   10.000000  12.000000   4.000000    8.000000   6.500000   
75%     7.750000   18.000000  24.000000   9.750000   23.750000  15.250000   
max    40.000000   60.000000  38.000000  47.000000   47.000000  38.000000   

          yrspain    yrssand     yrsweld  
count  272.000000  88.000000  290.000000  
mean    10.448529  10.920455   14.420690  
std     11.427822  10.875705   12.957834  
min      1.000000   1.000000    1.000000  
25%      2.000000   2.000000    3.000000  
50%      5.000000   7.500000   10.000000  
75%     15.000000  18.000000   25.000000  
max     50.000000  50.000000   50.000000  

[8 rows x 42 columns]

disease 32


ageadas    ageasbe     agebron     agechas     agechro     agecopd  \
count  344.000000  56.000000  151.000000  164.000000  511.000000  339.000000   
mean    49.976744  54.642857   38.357616    6.646341   40.927593   56.982301   
std     12.838741   7.354476   19.412829    3.939331   17.123446    7.469443   
min     13.000000  27.000000    1.000000    0.000000    1.000000   30.000000   
25%     42.750000  50.000000   21.000000    4.000000   30.000000   52.500000   
50%     53.000000  56.500000   40.000000    6.000000   45.000000   57.000000   
75%     60.000000  60.000000   55.000000    9.000000   55.000000   62.000000   
max     72.000000  65.000000   73.000000   17.000000   72.000000   73.000000   

          agediab     ageemph    agefibr     agehear  ...     diagdiab  \
count  504.000000  513.000000  15.000000  677.000000  ...  5276.000000   
mean    54.438492   56.116959  50.866667   54.355982  ...     0.098180   
std      9.963561    8.926455   9.804275   10.431178  ...     0.297587   
min      1.000000   13.000000  25.000000    1.000000  ...     0.000000   
25%     50.000000   51.000000  49.000000   50.000000  ...     0.000000   
50%     56.000000   57.000000  51.000000   56.000000  ...     0.000000   
75%     60.000000   62.000000  56.000000   61.000000  ...     0.000000   
max     73.000000   73.000000  65.000000   74.000000  ...     1.000000   

          diagemph     diagfibr     diaghear     diaghype     diagpneu  \
count  5269.000000  5273.000000  5262.000000  5272.000000  5271.000000   
mean      0.103625     0.003034     0.134170     0.355842     0.235249   
std       0.304802     0.055006     0.340867     0.478813     0.424195   
min       0.000000     0.000000     0.000000     0.000000     0.000000   
25%       0.000000     0.000000     0.000000     0.000000     0.000000   
50%       0.000000     0.000000     0.000000     0.000000     0.000000   
75%       0.000000     0.000000     0.000000     1.000000     0.000000   
max       1.000000     1.000000     1.000000     1.000000     1.000000   

          diagsarc     diagsili     diagstro     diagtube  
count  5269.000000  5269.000000  5276.000000  5276.000000  
mean      0.002467     0.001708     0.028810     0.011372  
std       0.049615     0.041298     0.167287     0.106043  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000  

[8 rows x 32 columns]

personal_cancer 30


ageblad    agebrea    agecerv    agecolo    ageesop    agekidn  \
count  35.000000  91.000000  78.000000  25.000000   5.000000   8.000000   
mean   55.885714  49.626374  36.538462  53.840000  50.800000  54.625000   
std    10.303544   8.452019  10.079801   8.979978   6.978539   7.424621   
min    17.000000  25.000000  21.000000  32.000000  45.000000  48.000000   
25%    52.000000  43.500000  28.000000  50.000000  46.000000  49.500000   
50%    56.000000  50.000000  34.500000  55.000000  48.000000  52.000000   
75%    63.500000  55.500000  41.750000  59.000000  53.000000  57.500000   
max    68.000000  68.000000  63.000000  67.000000  62.000000  67.000000   

         agelary  agelung    agenasa    ageoral  ...     canckidn  \
count   5.000000     4.00   3.000000  14.000000  ...  5276.000000   
mean   55.000000    65.25  62.000000  53.928571  ...     0.001516   
std    12.629331     4.50   6.244998   9.988180  ...     0.038914   
min    40.000000    59.00  55.000000  36.000000  ...     0.000000   
25%    43.000000    63.50  59.500000  50.000000  ...     0.000000   
50%    60.000000    66.50  64.000000  54.000000  ...     0.000000   
75%    65.000000    68.25  65.500000  60.750000  ...     0.000000   
max    67.000000    69.00  67.000000  70.000000  ...     1.000000   

          canclary     canclung     cancnasa     cancoral  cancpanc  \
count  5276.000000  5276.000000  5274.000000  5275.000000    5276.0   
mean      0.000948     0.000758     0.000569     0.002654       0.0   
std       0.030773     0.027527     0.023846     0.051454       0.0   
min       0.000000     0.000000     0.000000     0.000000       0.0   
25%       0.000000     0.000000     0.000000     0.000000       0.0   
50%       0.000000     0.000000     0.000000     0.000000       0.0   
75%       0.000000     0.000000     0.000000     0.000000       0.0   
max       1.000000     1.000000     1.000000     1.000000       0.0   

          cancphar     cancstom     cancthyr     canctran  
count  5276.000000  5276.000000  5273.000000  5272.000000  
mean      0.000190     0.001137     0.001517     0.000190  
std       0.013767     0.033707     0.038925     0.013772  
min       0.000000     0.000000     0.000000     0.000000  
25%       0.000000     0.000000     0.000000     0.000000  
50%       0.000000     0.000000     0.000000     0.000000  
75%       0.000000     0.000000     0.000000     0.000000  
max       1.000000     1.000000     1.000000     1.000000  

[8 rows x 30 columns]

family 5


fambrother    famsister     famchild    famfather    fammother
count  5140.000000  5141.000000  5138.000000  5114.000000  5152.000000
mean      0.057977     0.042988     0.004866     0.109894     0.065800
std       0.233722     0.202849     0.069591     0.312789     0.247956
min       0.000000     0.000000     0.000000     0.000000     0.000000
25%       0.000000     0.000000     0.000000     0.000000     0.000000
50%       0.000000     0.000000     0.000000     0.000000     0.000000
75%       0.000000     0.000000     0.000000     0.000000     0.000000
max       1.000000     1.000000     1.000000     1.000000     1.000000

alcohol 10


lss_alcohol_freq  lss_alcohol_num  acrin_alc_curr  acrin_alc_ever  \
count       3495.000000      2744.000000     1662.000000     1757.000000   
mean           2.931903         1.910350        1.737064        1.948776   
std            1.459018         0.912094        0.440361        0.220516   
min            1.000000         1.000000        1.000000        1.000000   
25%            2.000000         1.000000        1.000000        2.000000   
50%            3.000000         2.000000        2.000000        2.000000   
75%            4.000000         2.000000        2.000000        2.000000   
max            5.000000         5.000000        2.000000        2.000000   

       acrin_drink24hr  acrin_drinknum_curr  acrin_drinknum_form  \
count      1185.000000          1181.000000           404.000000   
mean          0.972152             6.938188            13.975248   
std           1.622811             9.341564            20.147949   
min           0.000000             0.000000             0.000000   
25%           0.000000             0.000000             0.000000   
50%           0.000000             4.000000             5.000000   
75%           2.000000            10.000000            20.000000   
max          26.000000            89.000000           100.000000   

       acrin_drinkyrs_curr  acrin_drinkyrs_form  acrin_lastdrink  
count          1174.000000           405.000000       435.000000  
mean             39.326235            23.711111         2.372414  
std              10.454303            13.555825         0.860914  
min               0.000000             0.000000         1.000000  
25%              35.000000            12.000000         1.000000  
50%              40.000000            23.000000         3.000000  
75%              45.000000            34.000000         3.000000  
max              70.000000            70.000000         3.000000

In [146]:
nlst_demos = nlst_dict_subset[allcols]
nlst_demos.describe()

pid        study          age       educat       ethnic  \
count    5282.000000  5282.000000  5282.000000  5282.000000  5282.000000   
mean   148184.307081     1.474252    62.183453     4.835290     2.538054   
std     44395.945951     0.728020     5.184738     4.579593     7.283255   
min    100012.000000     1.000000    54.000000     1.000000     1.000000   
25%    113113.500000     1.000000    58.000000     3.000000     2.000000   
50%    125989.000000     1.000000    61.000000     5.000000     2.000000   
75%    205225.250000     2.000000    66.000000     6.000000     2.000000   
max    218887.000000     3.000000    74.000000    99.000000    99.000000   

            gender       height      marital         race       weight  ...  \
count  5282.000000  5273.000000  5281.000000  5282.000000  5270.000000  ...   
mean      1.430140    67.846008     2.654043     1.307459   181.264706  ...   
std       0.495142     3.921241     1.247796     3.831637    38.917809  ...   
min       1.000000    49.000000     1.000000     1.000000    75.000000  ...   
25%       1.000000    65.000000     2.000000     1.000000   154.000000  ...   
50%       1.000000    68.000000     2.000000     1.000000   179.000000  ...   
75%       2.000000    71.000000     3.000000     1.000000   205.000000  ...   
max       2.000000    83.000000     9.000000    99.000000   368.000000  ...   

       lss_alcohol_freq  lss_alcohol_num  acrin_alc_curr  acrin_alc_ever  \
count       3495.000000      2744.000000     1662.000000     1757.000000   
mean           2.931903         1.910350        1.737064        1.948776   
std            1.459018         0.912094        0.440361        0.220516   
min            1.000000         1.000000        1.000000        1.000000   
25%            2.000000         1.000000        1.000000        2.000000   
50%            3.000000         2.000000        2.000000        2.000000   
75%            4.000000         2.000000        2.000000        2.000000   
max            5.000000         5.000000        2.000000        2.000000   

       acrin_drink24hr  acrin_drinknum_curr  acrin_drinknum_form  \
count      1185.000000          1181.000000           404.000000   
mean          0.972152             6.938188            13.975248   
std           1.622811             9.341564            20.147949   
min           0.000000             0.000000             0.000000   
25%           0.000000             0.000000             0.000000   
50%           0.000000             4.000000             5.000000   
75%           2.000000            10.000000            20.000000   
max          26.000000            89.000000           100.000000   

       acrin_drinkyrs_curr  acrin_drinkyrs_form  acrin_lastdrink  
count          1174.000000           405.000000       435.000000  
mean             39.326235            23.711111         2.372414  
std              10.454303            13.555825         0.860914  
min               0.000000             0.000000         1.000000  
25%              35.000000            12.000000         1.000000  
50%              40.000000            23.000000         3.000000  
75%              45.000000            34.000000         3.000000  
max              70.000000            70.000000         3.000000  

[8 rows x 140 columns]